src/metabase/notification/send.clj

```diff
@@ -188,12 +181,89 @@
       (prometheus/observe! :metabase-notification/total-duration-ms {:payload-type payload_type} total-time))
     nil))
 
+(defprotocol NotificationQueueProtocol
+  "Protocol for notification queue implementations."
+  (put-notification!  [this notification] "Add a notification to the queue. If a notification with the same id is already in the queue, replace it.")
+  (take-notification! [this]              "Take the next notification from the queue, blocking if none available."))
+
+(deftype ^:private NotificationQueue
+         [^java.util.LinkedList ids-list
+          ^java.util.concurrent.ConcurrentHashMap id->notification
+          ^java.util.concurrent.locks.ReentrantLock queue-lock
+          ^java.util.concurrent.locks.Condition not-empty-cond]
+  NotificationQueueProtocol
+  (put-notification! [_ notification]
+    (let [id (or (:id notification) (str (random-uuid)))]
+      (.lock queue-lock)
+      (try
+        (when-not (.contains ids-list id)
```

Comment:
Let's avoid this linear scan by rather checking whether its a key in `id->notification`.